cloudinit/distros/__init__.py

```diff
@@ -139,9 +157,88 @@ def _unpickle(self, ci_pkl_version: int) -> None:
+        """
+        packages_by_manager = defaultdict(set)
+        generic_packages: Set[str] = set()
+        for entry in pkglist:
+            if isinstance(entry, dict):
+                for package_manager, package_list in entry.items():
+                    for definition in package_list:
+                        if isinstance(definition, list):
+                            definition = tuple(definition)
+                        try:
+                            packages_by_manager[
+                                known_package_managers[package_manager]
+                            ].add(definition)
+                        except KeyError:
+                            LOG.error(
+                                "Cannot install packages under %s as it is "
+                                "not a supported package manager!",
+                                package_manager,
+                            )
+            elif isinstance(entry, str):
+                generic_packages.add(entry)
+            else:
+                raise ValueError(
+                    "Invalid 'packages' yaml specification. "
+                    "Check schema definition."
+                )
+        return dict(packages_by_manager), generic_packages
+
     def install_packages(self, pkglist):
-        raise NotImplementedError()
+        error_message = (
+            "Failed to install the following packages: %s. "
+            "See associated package manager logs for more details."
+        )
+        # If an entry hasn't been included with an explicit package name,
+        # add it to a 'generic' list of packages
+        (
+            packages_by_manager,
+            generic_packages,
+        ) = self._extract_package_by_manager(pkglist)
+
+        # First install packages using package manager(s)
+        # supported by the distro
+        uninstalled = []
+        for manager in self.package_managers:
+            to_try = (
+                packages_by_manager.get(manager.__class__, set())
+                | generic_packages
+            )
+            uninstalled = manager.install_packages(to_try)
+            failed = {
+                pkg for pkg in uninstalled if pkg not in generic_packages
+            }
+            if failed:
+                LOG.error(error_message, failed)
+            generic_packages = set(uninstalled)
+
+        # Now attempt any specified package managers not explicitly supported
+        # by distro
+        for manager, packages in packages_by_manager.items():
+            if manager.name not in known_package_managers:
+                LOG.error(
+                    "Cannot install any packages listed under unknown package "
+                    "manager: %s.",
+                    manager,
+                )
+            if manager.name in [p.name for p in self.package_managers]:
+                # We already installed/attempt these; don't try again
+                continue
+            uninstalled.extend(
```

Comment:
1. Nothing is covering this in tests either. Maybe we want something like:

```diff
--- a/tests/unittests/config/test_cc_package_update_upgrade_install.py
+++ b/tests/unittests/config/test_cc_package_update_upgrade_install.py
@@ -75,6 +75,17 @@ class TestMultiplePackageManagers:
         assert mock.call(["snap", "install", "pkg1"]) in m_subp.call_args_list
         assert mock.call(["snap", "install", "pkg2"]) in m_subp.call_args_list
 
+    @mock.patch("cloudinit.subp.subp")
+    def test_explicit_snap_when_not_distro_default(self, m_subp, common_mocks):
+        import pdb; pdb.set_trace()
+        cloud = get_cloud("debian")
+        cfg = {"packages": [{"snap": ["pkg1", "pkg2"]}]}
+        handle("", cfg, cloud, [])
+
+        assert len(m_subp.call_args_list) == 2
+        assert mock.call(["snap", "install", "pkg1"]) in m_subp.call_args_list
+        assert mock.call(["snap", "install", "pkg2"]) in m_subp.call_args_list
+
     @mock.patch("cloudinit.subp.subp")
     def test_explicit_snap_version(self, m_subp, common_mocks):
         cloud = get_cloud("ubuntu")
```


2. One thought this brings up is, if debian doesn't have snap command installed should the package_manager have a `available/viable` check to determine if it is even compatible? If not we will traceback with a called process error here on `command not found`. We can either add a `PackageManager.is_applicable/available` check or better handle a ProcessExecutionError when the PackageManager primary command is absent.

Reply:
Per your 2, I think I just missed a `continue`. If you go up two blocks, we have
```
            if manager.name not in known_package_managers:
                LOG.error(
                    "Cannot install any packages listed under unknown package "
                    "manager: %s.",
                    manager,
                )
```
That was meant to cover the case where somebody specifies something like snap but snap isn't installed. If I add continue after the error, that should address the problem,correct?

Reply:
(original comment deleted).

This will be refactored in the next push